src/hotspot/share/nmt/vmatree.cpp

```diff
@@ -0,0 +1,206 @@
+    int cmp_B = AddressComparator::cmp(head->key(), B);
+    stB.out = head->val().out;
+    if (cmp_B < 0) {
+      // Record all nodes preceding B.
+      to_be_deleted_inbetween_a_b.push({head->key(), head->val()});
+    } else if (cmp_B == 0) {
+      // Re-purpose B node, unless it would result in a noop node, in
+      // which case record old node at B for deletion and summary accounting.
+      if (stB.is_noop()) {
+        to_be_deleted_inbetween_a_b.push(AddressState{B, head->val()});
+      } else {
+        head->val() = stB;
+      }
+      B_needs_insert = false;
+    }
+  });
+
+  // Insert B node if needed
+  if (B_needs_insert && // Was not already inserted
+      !stB.is_noop())   // The operation is differing
+    {
+    _tree.upsert(B, stB);
+  }
+
+  // We now need to:
+  // a) Delete all nodes between (A, B]. Including B in the case of a noop.
+  // b) Perform summary accounting
+  SummaryDiff diff;
+
+  if (to_be_deleted_inbetween_a_b.length() == 0 && LEQ_A_found) {
+    // We must have smashed a hole in an existing region (or replaced it entirely).
+    // LEQ_A - A - B - (some node >= B)
+    auto& rescom = diff.flag[NMTUtil::flag_to_index(LEQ_A.out().flag())];
+    if (LEQ_A.out().type() == StateType::Reserved) {
+      rescom.reserve -= B - A;
+    } else if (LEQ_A.out().type() == StateType::Committed) {
+      rescom.commit -= B - A;
+      rescom.reserve -= B - A;
+    }
+  }
+
+  // Track the previous node.
+  AddressState prev{A, stA};
+  for (int i = 0; i < to_be_deleted_inbetween_a_b.length(); i++) {
+    const AddressState delete_me = to_be_deleted_inbetween_a_b.at(i);
+    _tree.remove(delete_me.address);
+
+    // Perform summary accounting
+    auto& rescom = diff.flag[NMTUtil::flag_to_index(delete_me.in().flag())];
+    if (delete_me.in().type() == StateType::Reserved) {
+      rescom.reserve -= delete_me.address - prev.address;
+    } else if (delete_me.in().type() == StateType::Committed) {
+      rescom.commit -= delete_me.address - prev.address;
+      rescom.reserve -= delete_me.address - prev.address;
+    }
+    prev = delete_me;
+  }
+
+  if (prev.address != A && prev.out().type() != StateType::Released) {
+    // The last node wasn't released, so it must be connected to a node outside of (A, B)
+    // A - prev - B - (some node >= B)
+    // It might be that prev.address == B == (some node >= B), this is fine.
+    if (prev.out().type() == StateType::Reserved) {
+      auto& rescom = diff.flag[NMTUtil::flag_to_index(prev.out().flag())];
+      rescom.reserve -= B - prev.address;
+    } else if (prev.out().type() == StateType::Committed) {
+      auto& rescom = diff.flag[NMTUtil::flag_to_index(prev.out().flag())];
+      rescom.commit -= B - prev.address;
+      rescom.reserve -= B - prev.address;
+    }
+  }
+
+  // Finally, we can register the new region [A, B)'s summary data.
+  auto& rescom = diff.flag[NMTUtil::flag_to_index(metadata.flag)];
```

Comment:
Can we please not use `auto` when unnecessary? IDEs have enough of a hard time understanding our code. `SingleDiff&`, please.